server/pkg/api/pool.go

```diff
@@ -0,0 +1,256 @@
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"math/rand"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	lxd "github.com/lxc/lxd/client"
+	"github.com/lxc/lxd/shared/api"
+
+	"github.com/whywaita/shoes-lxd-multi/server/pkg/lxdclient"
+	"github.com/whywaita/shoes-lxd-multi/server/pkg/metric"
+)
+
+type gotInstances struct {
+	Instances         []api.Instance
+	OverCommitPercent uint64
+	Error             error
+}
+
+func getInstancesWithTimeout(_ctx context.Context, h lxdclient.LXDHost, d time.Duration, l *slog.Logger) ([]api.Instance, uint64, error) {
+	ret := make(chan *gotInstances)
+	ctx, cancel := context.WithTimeout(_ctx, d)
+	defer cancel()
+	go func() {
+		defer close(ret)
+		s, err := lxdclient.GetAnyInstances(h.Client)
+		if err != nil {
+			ret <- &gotInstances{
+				Instances:         nil,
+				OverCommitPercent: 0,
+				Error:             fmt.Errorf("failed to get instances: %w", err),
+			}
+			return
+		}
+		r, err := lxdclient.GetResource(ctx, h.HostConfig, l)
+		if err != nil {
+			ret <- &gotInstances{
+				Instances:         nil,
+				OverCommitPercent: 0,
+				Error:             fmt.Errorf("failed to get resource: %w", err),
+			}
+			return
+		}
+		var used uint64
+		for _, i := range s {
+			if i.StatusCode != api.Running {
+				continue
+			}
+			instanceCPU := i.Config["limits.cpu"]
+			if instanceCPU == "" {
+				continue
+			}
+			cpu, err := strconv.Atoi(i.Config["limits.cpu"])
+			if err != nil {
+				ret <- &gotInstances{
+					Instances:         nil,
+					OverCommitPercent: 0,
+					Error:             fmt.Errorf("failed to parse limits.cpu: %w", err),
+				}
+				return
+			}
+			used += uint64(cpu)
+		}
+		overCommitPercent := uint64(float64(used) / float64(r.CPUTotal) * 100)
+		ret <- &gotInstances{
+			Instances:         s,
+			OverCommitPercent: overCommitPercent,
+			Error:             nil,
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, 0, errors.New("timed out")
+		case r := <-ret:
+			return r.Instances, r.OverCommitPercent, r.Error
+		}
+	}
+}
+
+type instance struct {
+	Host         *lxdclient.LXDHost
+	InstanceName string
+}
+
+func findInstances(ctx context.Context, targets []lxdclient.LXDHost, match func(api.Instance) bool, limitOverCommit uint64, l *slog.Logger) []instance {
+	type result struct {
+		host              *lxdclient.LXDHost
+		overCommitPercent uint64
+		instances         []string
+	}
+	rs := make([]result, len(targets))
+
+	wg := new(sync.WaitGroup)
+	wg.Add(len(targets))
```

Comment:
Could you move to inside of range? ( to `wg.Add(1)` )
Maybe this occurs in infinite waiting if failure.